packages/react-storage/src/components/StorageBrowser/views/hooks/useSearch.ts

```diff
@@ -0,0 +1,77 @@
+import React from 'react';
+import debounce from 'lodash/debounce';
+
+const DEFAULT_DELAY = 300;
+export interface InitialSearchValues<T> {
+  searchKey: keyof T;
+  searchTerm?: string;
+  debounceDelay?: number;
+}
+
+interface UseSearchProps<T> {
+  initialValues: InitialSearchValues<T>;
+  items: T[];
+  onSearch: (term: string, includeSubfolders?: boolean) => void;
+}
+
+interface UseSearchResult<T> {
+  filteredItems: T[];
+  searchTerm: string;
+  onSearch: (term: string, includeSubfolders?: boolean) => void;
+}
+
+/**
+ * Custom hook to search a list of items based on a search key and search term.
+ * @template T - The type of items in the search list.
+ * @param {UseSearchProps<T>} props
+ * @returns {UseSearch<T>} - The search results, current term, and handler.
+ */
+export function useSearch<T>({
+  items,
+  initialValues: initialValuesProp,
+  onSearch,
+}: UseSearchProps<T>): UseSearchResult<T> {
+  const [initialValues] = React.useState(initialValuesProp);
+  const {
+    searchTerm: initialSearchTerm = '',
+    searchKey,
+    debounceDelay = DEFAULT_DELAY,
+  } = initialValues;
+
+  const [searchTerm, setSearchTerm] = React.useState(initialSearchTerm);
+
+  const filteredItems = React.useMemo(
+    () =>
+      items.filter((item) => {
+        const test = item[searchKey];
+        if (typeof test === 'string') {
+          return test.includes(searchTerm);
+        }
+      }),
+    [items, searchKey, searchTerm]
+  );
```

Comment:
Can we handle this logic in the list handler so we don't have to track `searchTerm` in state?

Reply:
I think it's better to handle it in useSearch because that allows us to consolidate logic for regular search and "subfolder" search. We don't always call the list handler.

Reply:
Another option is to always call list handlers on search, and use previous state to search without subfolders. But we'll couple search with data fetching.